File: leetcode-100-Same_Tree/leetcode-100-Same_Tree.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        if not p and not q:
            return True
        if not p or not q or p.val != q.val:
            return False
        v1 = self.isSameTree(p.left, q.left)
        v2 = self.isSameTree(p.right, q.right)
        return True if v1 and v2 else False
    

if __name__ == "__main__":
    # p: TreeNode = TreeNode(1, TreeNode(2), TreeNode(3))
    # q: TreeNode = TreeNode(1, TreeNode(2), TreeNode(3))

    # p: TreeNode = TreeNode(1, TreeNode(2))
    # q: TreeNode = TreeNode(1, None, TreeNode(2))

    p: TreeNode = TreeNode(1, TreeNode(2), TreeNode(1))
    q: TreeNode = TreeNode(1, TreeNode(1), TreeNode(2))

    res: bool = Solution().isSameTree(p, q)

    print(res)

File: leetcode-100-Same_Tree/leetcode_100_Same_Tree.java
class leetcode_100_Same_Tree
{
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    
    public boolean isSameTree(TreeNode p, TreeNode q) {
        if(p == null && q == null)
        {
            return true;
        }
        if((p == null || q == null) || p.val != q.val)
        {
            return false;
        }
        boolean b1 = isSameTree(p.left, q.left),
                b2 = isSameTree(p.right, q.right);
        return (b1 && b2);
    }
    
    public static void main(String[] args)
    {
        TreeNode p = new TreeNode(1, new TreeNode(2), new TreeNode(3)),
                 q = new TreeNode(1, new TreeNode(2), new TreeNode(3));
        boolean result =  new leetcode_100_Same_Tree().isSameTree(p, q);
        System.out.println(result);
    }
}


 
